pkg/cmd/roachtest/tests/mixed_version_c2c.go

```diff
@@ -0,0 +1,442 @@
+)
+
+func runC2CMixedVersions(ctx context.Context, t test.Test, c cluster.Cluster, sp replicationSpec) {
+	cm := InitC2CMixed(ctx, t, c, sp)
+	cm.SetupHook(ctx)
+	cm.WorkloadHook(ctx)
+	cm.LatencyHook(ctx)
+	cm.UpdateHook(ctx)
+	cm.Run(ctx, c)
+}
+
+func InitC2CMixed(
+	ctx context.Context, t test.Test, c cluster.Cluster, sp replicationSpec,
+) *c2cMixed {
+	// TODO(msbutler): allow for version skipping and multiple upgrades.
+	sourceMvt := mixedversion.NewTest(ctx, t, t.L(), c, c.Range(1, sp.srcNodes),
+		mixedversion.MinimumSupportedVersion(minSupportedVersion),
+		mixedversion.AlwaysUseLatestPredecessors,
+		mixedversion.NumUpgrades(expectedMajorUpgrades),
+		mixedversion.EnabledDeploymentModes(mixedversion.SharedProcessDeployment),
+		mixedversion.WithTag("source"),
+		mixedversion.DisableSkipVersionUpgrades,
+	)
+
+	destMvt := mixedversion.NewTest(ctx, t, t.L(), c, c.Range(sp.srcNodes+1, sp.srcNodes+sp.dstNodes),
+		mixedversion.MinimumSupportedVersion(minSupportedVersion),
+		mixedversion.AlwaysUseLatestPredecessors,
+		mixedversion.NumUpgrades(expectedMajorUpgrades),
+		mixedversion.EnabledDeploymentModes(mixedversion.SystemOnlyDeployment),
+		mixedversion.WithTag("dest"),
+		mixedversion.DisableSkipVersionUpgrades,
+	)
+
+	return &c2cMixed{
+		sourceMvt:           sourceMvt,
+		destMvt:             destMvt,
+		sourceStartedChan:   make(chan struct{}),
+		destStartedChan:     make(chan struct{}),
+		sp:                  sp,
+		t:                   t,
+		c:                   c,
+		fingerprintArgsChan: make(chan fingerprintArgs, 1),
+		fingerprintChan:     make(chan int64, 1),
+	}
+}
+
+type sourceTenantInfo struct {
+	name  string
+	pgurl *url.URL
+}
+
+type fingerprintArgs struct {
+	retainedTime hlc.Timestamp
+	cutoverTime  hlc.Timestamp
+}
+
+type c2cMixed struct {
+	sourceMvt, destMvt  *mixedversion.Test
+	sourceStartedChan   chan struct{}
+	destStartedChan     chan struct{}
+	sp                  replicationSpec
+	t                   test.Test
+	c                   cluster.Cluster
+	fingerprintArgsChan chan fingerprintArgs
+	fingerprintChan     chan int64
+	// midUpgradeCatchupMu attempts to prevent the source from upgrading while the
+	// destination is waiting for the stream to catch up in some mixed version
+	// state.
+	midUpgradeCatchupMu syncutil.Mutex
+
+	ingestionJobID  catpb.JobID
+	workloadStopper mixedversion.StopFunc
+}
+
+func (cm *c2cMixed) SetupHook(ctx context.Context) {
+	sourceChan := make(chan sourceTenantInfo, 1)
+
+	cm.sourceMvt.OnStartup(
+		"generate pgurl",
+		func(ctx context.Context, l *logger.Logger, r *rand.Rand, h *mixedversion.Helper) error {
+			// Enable rangefeeds, required for PCR to work.
+			l.Printf("enabling rangefeeds")
+			if err := h.System.Exec(r, "SET CLUSTER SETTING kv.rangefeed.enabled = true"); err != nil {
+				return errors.Wrap(err, "failed to enable rangefeeds")
+			}
+			close(cm.sourceStartedChan)
```

Comment:
I think we might be served well by a comment on why this is the right place to close this channel.